enhancements/cert-manager-proxy/http01_challenge_proxy.md

```diff
@@ -0,0 +1,364 @@
+     - If the API certificate has expired, use insecure connection flags to access the cluster and perform remediation.
+     - Monitor for certificate expiry and configure alerts to notify administrators before expiry occurs.
+   - **Detection**: Configure alerts for expiring certificates. Warning alerts should be triggered when certificates are close to expiry, and critical alerts when certificates have expired. This allows administrators to take action before cluster API access is impacted.
+
+2. **Traffic Interference**: The proxy could inadvertently interfere with other traffic. Mitigation: Carefully scope the proxy's functionality to only handle HTTP01 challenge traffic.
+
+### Implementation History
+
+- **2025-03-28**: Enhancement proposal created.
+
+### References
+
+- [Cert Manager Expansion JIRA Epic](https://issues.redhat.com/browse/CNF-13731)
+- [ACME HTTP01 Challenge](https://letsencrypt.org/docs/challenge-types/#http-01-challenge)
+- [Proxy Code Repository](https://github.com/mvazquezc/cert-mgr-http01-proxy/tree/main)
+- [Deployment Manifest](https://github.com/mvazquezc/cert-mgr-http01-proxy/blob/main/manifests/deploy-in-ocp.yaml)
+
+### Workflow Description
+
+1. Cert Manager initiates an HTTP01 challenge for the API endpoint (`api.cluster.example.com`).
+2. The HTTP01 challenge request is directed to the API VIP on port 80 with the path `/.well-known/acme-challenge/<token>`.
+3. **Traffic Identification**: The HTTP01 Challenge Proxy identifies this as ACME challenge traffic by examining the destination path (`/.well-known/acme-challenge/*`) rather than using SNI validation.
+4. **Selective Redirection**: Only traffic matching the ACME challenge path pattern is intercepted using `nftables` and redirected to the proxy pod on port 8888. Other HTTP requests to the API VIP on port 80 receive a **403 Forbidden** response.
+5. The proxy pod forwards the filtered request to the OpenShift Ingress Router, which serves the challenge response from the Cert Manager challenge pod.
+6. The ACME CA validates the challenge and issues the certificate for the API endpoint.
+
+### Topology Considerations
+
+- **Standard Clusters**: The API VIP is hosted on control plane nodes. The proxy ensures that HTTP01 challenges are redirected to the OpenShift Ingress Routers.
+- **Compact Clusters**: The proxy handles scenarios where the API VIP node may or may not host an OpenShift Router, ensuring consistent challenge redirection.
+- **SNO (Single Node OpenShift)**: The proxy is not strictly required in this topology, as the API and wildcard FQDNs resolve to the same IP. However, it can still be deployed for consistency.
+
+#### Hypershift / Hosted Control Planes
+
+This enhancement does not directly apply to Hypershift deployments, as the API endpoint management in Hypershift differs from baremetal environments. However, the proxy's design could be adapted for similar use cases in Hypershift if needed.
```

Comment:
Because you are making changes to the `apiservers.config.openshift.io` resource this _does_ impact HyperShift as HyperShift's APIs embed the spec for that resource in their resources: https://github.com/openshift/hypershift/blob/d5884761640027a4d4d78fe1711597a9cef867a7/api/hypershift/v1beta1/hostedcluster_types.go#L1822-L1826

If you do not want this feature to apply to HyperShift, how do you intend to prevent the field from being embedded within HyperShift types (or how do you prevent HyperShift users from being able to set it if they can't use it)?

Reply:
We haven't tested in HyperShift, and I don't think the solution will work as it is today on HyperShift since it relies on MCO  that works differently in HyperShift.

Does the KASO know the topology it is running on? in case it does, we can have it ignore this setting in HyperShift topology until we decide how this gets implemented in HyperShift. WDYT?

Reply:
Pinged @sjenning @csrwng and @enxebre on Slack to comment here.